Added Wasmer Js API

syrusakbary opened the pull request:
<!-- 
Prior to submitting a PR, review the CONTRIBUTING.md document for recommendations on how to test:
https://github.com/wasmerio/wasmer/blob/master/CONTRIBUTING.md#pull-requests

-->

# Description

This PR adds support for compiling Wasmer into JS/Wasm via wasm-bindgen, so 3rd party dependencies can use Wasmer to target the web also.

Things working:
* [x] Type introspection: we have to decide how to do it, either run our own parser on the data or rely on the future APIs for introspect Wasm types (this APIs are yet available in the browser, but a proposal is on the way: [wasm-js-types proposal](https://github.com/WebAssembly/js-types)).
* [x] Wasm-pack wasmer-js tests
* [x] Assert 0 warnings in lint
* [x] Allow for custom JS errors (via `RuntimeError`)
* Externals
   * [x] Functions with and without Environments
      * [x] Native functions
      * [x] Dynamic functions
   * [x] Memory
   * [x] Globals
   * [x] Tables

# Review


- [ ] Add a short description of the change to the CHANGELOG.md file

Hywan reviewed (commented):
First round of review.

syrusakbary commented on .github/workflows/js.yaml:
```diff
@@ -0,0 +1,27 @@
+      - '**'
+      - '!master'
+
+name: wasmerjs
```
> I'm saying that to avoid confusion with wasmer-js, https://github.com/wasmerio/wasmer-js/.

This is indended to replace `wasmer-js`. That's why I chose this name :)

syrusakbary replied:
This is the job name, no need to change it as I'm not sure Github allows `-` inside job names

syrusakbary commented on lib/js-api/src/env.rs:
```diff
@@ -0,0 +1,228 @@
+}
+
+unsafe impl<T: Send> Send for LazyInit<T> {}
+// I thought we could opt out of sync..., look into this
+// unsafe impl<T> !Sync for InitWithInstance<T> {}
```
This is copied from wasmer api. No need to resolve here :)

syrusakbary commented on lib/js-api/src/exports.rs:
```diff
@@ -0,0 +1,324 @@
+/// Exports is a special kind of map that allows easily unwrapping
+/// the types of instances.
+///
+/// TODO: add examples of using exports
```
This is the same as Wasmer API

syrusakbary commented on lib/js-api/src/exports.rs:
```diff
@@ -0,0 +1,324 @@
+where
+    I: Iterator<Item = (&'a String, &'a Extern)> + Sized,
+{
+    // /// Get only the functions.
+    // pub fn functions(self) -> impl Iterator<Item = (&'a String, &'a Function)> + Sized {
+    //     self.iter.filter_map(|(name, export)| match export {
+    //         Extern::Function(function) => Some((name, function)),
+    //         _ => None,
+    //     })
+    // }
```
This should be fixed. Somehow the diff is outdated

syrusakbary commented on lib/js-api/src/exports.rs:
```diff
@@ -0,0 +1,324 @@
+        })
+    }
+
+    // /// Get only the globals.
+    // pub fn globals(self) -> impl Iterator<Item = (&'a String, &'a Global)> + Sized {
+    //     self.iter.filter_map(|(name, export)| match export {
+    //         Extern::Global(global) => Some((name, global)),
+    //         _ => None,
+    //     })
+    // }
+
+    // /// Get only the tables.
+    // pub fn tables(self) -> impl Iterator<Item = (&'a String, &'a Table)> + Sized {
+    //     self.iter.filter_map(|(name, export)| match export {
+    //         Extern::Table(table) => Some((name, table)),
+    //         _ => None,
+    //     })
+    // }
```
This should be fixed. Somehow the diff is outdated

syrusakbary commented on lib/derive/src/lib.rs:
```diff
@@ -27,7 +27,7 @@ fn impl_wasmer_env_for_struct(
     let where_clause = generics.where_clause.clone();
     quote! {
-        impl < #lifetimes_and_generics > ::wasmer::WasmerEnv for #name < #lifetimes_and_generics > #where_clause{
+        impl < #lifetimes_and_generics > ::wasmer_js::WasmerEnv for #name < #lifetimes_and_generics > #where_clause{
```
Yup, I know. I intend to resolve it before merging (it just was a bit faster to deal at the end)

syrusakbary replied:
Fixed

syrusakbary commented on lib/js-api/src/export.rs:
```diff
@@ -0,0 +1,136 @@
+    Table(VMTable),
+
+    /// A memory export value.
+    Memory(VMMemory),
```
Not sure if I understand this change.

syrusakbary commented on lib/js-api/src/native.rs:
```diff
@@ -0,0 +1,269 @@
+);
+impl_native_traits!(
+    A1, A2, A3, A4, A5, A6, A7, A8, A9, A10, A11, A12, A13, A14, A15, A16, A17, A18, A19, A20
+);
```
This is already resolved in latest commits. Somehow your diff didn't got it :(

syrusakbary commented on lib/js-api/src/lib.rs:
```diff
@@ -0,0 +1,386 @@
+
+//     let expected = vec![Val::F64(5.0)].into_boxed_slice();
+//     exported.call(&[Val::I32(4)]) == Ok(expected)
+// }
```
I just wanted to test the Wasm size. I'll remove before merging

syrusakbary replied:
Removed

syrusakbary commented on Makefile:
```diff
@@ -480,6 +480,9 @@ test-packages:
 	cargo test --manifest-path lib/cli/Cargo.toml $(compiler_features) --release
 
+test-js:
+	cd lib/js-api && wasm-pack test --node -- --features=wat
```
wasm-pack can't handle doctests unfortunately

syrusakbary commented on lib/js-api/src/export.rs:
```diff
@@ -0,0 +1,136 @@
+    /// A memory export value.
+    Memory(VMMemory),
+    // /// A global export value.
+    // Global(VMGlobal),
```
`wasm-bindgen` has not been published yet (and it's not likely to be soon based on the input from Alex Chrichton)

syrusakbary commented on lib/js-api/src/exports.rs:
```diff
@@ -0,0 +1,324 @@
+use crate::export::Export;
+use crate::externals::{Extern, Function /* , Global, Table */, Memory};
```
This is already pushed in the PR. Somehow you are seeing an old diff :(

syrusakbary commented on lib/js-api/src/exports.rs:
```diff
@@ -0,0 +1,324 @@
+    // /// Get an export as a `Global`.
+    // pub fn get_global(&self, name: &str) -> Result<&Global, ExportError> {
+    //     self.get(name)
+    // }
```
This is already pushed in the PR. Somehow you are seeing an old diff :(

syrusakbary commented on lib/js-api/src/exports.rs:
```diff
@@ -0,0 +1,324 @@
+    //         .native()
+    //         .map_err(|_| ExportError::IncompatibleType)
+    // }
+
```
This is already pushed in the PR. Somehow you are seeing an old diff :(

syrusakbary commented on lib/js-api/src/lib.rs:
```diff
@@ -0,0 +1,386 @@
+pub use crate::exports::{ExportError, Exportable, Exports, ExportsIterator};
+pub use crate::externals::{
+    Extern, FromToNativeWasmType, Function, HostFunction, Memory, Table,
+    /* Global, */
```
This is already pushed in the PR. Somehow you are seeing an old diff :(

syrusakbary commented on lib/js-api/src/externals/function.rs:
```diff
@@ -0,0 +1,1620 @@
+        // let func_ref =
+        //     Val::into_vm_funcref(&Val::FuncRef(Some(self.clone())), &self.store).unwrap();
+        // std::ptr::write(p as *mut VMFuncRef, func_ref);
+        unimplemented!();
```
This is already pushed in the PR. Somehow you are seeing an old diff :(

syrusakbary commented on lib/js-api/src/externals/function.rs:
```diff
@@ -0,0 +1,1620 @@
+        //     // They should never be triggered, so we just panic.
+        //     Val::FuncRef(None) => panic!("Null funcref found in `Function::read_value_from`!"),
+        //     other => panic!("Invalid value in `Function::read_value_from`: {:?}", other),
+        // }
```
This is already pushed in the PR. Somehow you are seeing an old diff :(

syrusakbary commented on lib/js-api/src/externals/function.rs:
```diff
@@ -0,0 +1,1620 @@
+    //     let signature = self.ty();
+    //     if signature.params().len() != params.len() {
+    //         return Err(RuntimeError::new(format!(
+    //             "Parameters of type [{}] did not match signature {}",
```
This is already cleaned up. Somehow you are seeing an old diff :(

syrusakbary commented on lib/js-api/src/externals/global.rs:
```diff
@@ -0,0 +1,242 @@
+            _ => Err(ExportError::IncompatibleType),
+        }
+    }
+}
```
This is ported already into JS, it's tested and works properly.

syrusakbary commented on lib/js-api/src/lib.rs:
```diff
@@ -0,0 +1,386 @@
+}
+
+// #[cfg(test)]
+// wasm_bindgen_test::wasm_bindgen_test_configure!(run_in_browser);
```
Was just for testing.

syrusakbary commented on lib/js-api/src/import_object.rs:
````diff
@@ -0,0 +1,411 @@
+    ///
+    /// # Usage
+    /// ```ignore
+    /// # use wasmer_vm::{ImportObject, Instance, Namespace};
````
This is already cleaned up. Somehow you are seeing an old diff :(

syrusakbary commented on lib/js-api/src/externals/mod.rs:
```diff
@@ -0,0 +1,113 @@
+    /// A external [`Function`].
+    Function(Function),
+    // /// A external [`Global`].
+    // Global(Global),
```
This is already cleaned up. Somehow you are seeing an old diff :(

syrusakbary commented on lib/js-api/src/externals/memory.rs:
```diff
@@ -0,0 +1,322 @@
+        return bytes;
+        // let def = unsafe { definition.as_ref() };
+        // def.current_length.into()
+    }
```
This is already cleaned up. Somehow you are seeing an old diff :(

syrusakbary commented on lib/js-api/src/externals/function.rs:
```diff
@@ -0,0 +1,1620 @@
+///
+/// Spec: <https://webassembly.github.io/spec/core/exec/runtime.html#function-instances>
+///
+/// # Panics
+/// - Closures (functions with captured environments) are not currently supported
+///   with native functions. Attempting to create a native `Function` with one will
+///   result in a panic.
+///   [Closures as host functions tracking issue](https://github.com/wasmerio/wasmer/issues/1840)
```
Yup (for now)

syrusakbary commented on lib/js-api/src/exports.rs:
```diff
@@ -0,0 +1,324 @@
+    // /// Get an export as a `Table`.
+    // pub fn get_table(&self, name: &str) -> Result<&Table, ExportError> {
+    //     self.get(name)
+    // }
```
This is already cleaned up. Somehow you are seeing an old diff :(

syrusakbary commented:
@Hywan addressed all your comments except the ones related to using `wasmer-js-api` instead of `wasmer-js`.
I strongly believe `wasmer-js` is a much better name, it also servers as a clear replacement (intended) for https://github.com/wasmerio/wasmer-js, which is the end goal of this PR

syrusakbary commented:
bors try

syrusakbary commented on .github/workflows/js.yaml:
```diff
@@ -0,0 +1,27 @@
+  push:
+    branches:
+      - '**'
+      - '!master'
```
Fixed

syrusakbary commented on lib/api/src/cell.rs:
```diff
@@ -121,10 +120,10 @@ impl<'a, T: Copy> WasmCell<'a, T> {
     #[inline]
     fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
-        self.inner.fmt(f)
+        write!(f, "WasmCell({:?})", self.inner.get())
```
I was debugging WASI and this really helped

syrusakbary commented on lib/api/src/ptr.rs:
```diff
@@ -281,7 +281,13 @@ impl<T: Copy, Ty> Eq for WasmPtr<T, Ty> {}
 impl<T: Copy, Ty> fmt::Debug for WasmPtr<T, Ty> {
     fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
-        write!(f, "WasmPtr({:#x})", self.offset)
+        write!(
+            f,
+            "WasmPtr(offset: {}, pointer: {:#x}, align: {})",
+            self.offset,
+            self.offset,
+            mem::align_of::<T>()
+        )
```
I was debugging WASI and this really helped

syrusakbary commented:
bors try

syrusakbary commented on lib/js-api/README.md:
```diff
@@ -0,0 +1,63 @@
+
+## Config flags
+
+Wasmer has the following configuration flags:
```
```suggestion
`wasmer-js` has the following feature flags:
```

syrusakbary commented on lib/js-api/src/cell.rs:
````diff
@@ -0,0 +1,139 @@
+    ///
+    /// ```
+    /// use std::cell::Cell;
+    /// use wasmer::WasmCell;
````
```suggestion
    /// use wasmer_js::WasmCell;
```

syrusakbary commented:
bors try

syrusakbary commented:
bors try

Hywan commented on lib/js-api/src/externals/memory.rs:
```diff
@@ -0,0 +1,319 @@
+    Generic(String),
+}
+
+#[wasm_bindgen]
+extern "C" {
+    /// [MDN documentation](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/WebAssembly/Memory)
+    #[wasm_bindgen(js_namespace = WebAssembly, extends = js_sys::Object, typescript_type = "WebAssembly.Memory")]
+    #[derive(Clone, Debug, PartialEq, Eq)]
+    pub type JSMemory;
+
+    /// The `grow()` protoype method of the `Memory` object increases the
+    /// size of the memory instance by a specified number of WebAssembly
+    /// pages.
+    ///
+    /// Takes the number of pages to grow (64KiB in size) and returns the
+    /// previous size of memory, in pages.
+    ///
+    /// [MDN documentation](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/WebAssembly/Memory/grow)
+    #[wasm_bindgen(catch, method, js_namespace = WebAssembly)]
+    pub fn grow(this: &JSMemory, pages: u32) -> Result<u32, JsValue>;
+}
```
Already implemented, https://github.com/rustwasm/wasm-bindgen/blob/b97837b0989b58595f3ef72bd5691ce420591204/crates/js-sys/src/lib.rs#L3674-L3683. It's not marked as `catch` though. Is it the difference here? Is it on-purpose? If yes, then it must be documented here, and maybe open a PR to `wasm-bindgen`?

Hywan replied:
I've seen this https://github.com/rustwasm/wasm-bindgen/pull/2599, so it's fine. It must be written in the documentation. I'm adding a suggestion above.

syrusakbary replied:
Yeah, that's the difference.

Hywan commented on lib/js-api/src/externals/memory.rs:
```diff
@@ -0,0 +1,319 @@
+    /// pages.
+    ///
+    /// Takes the number of pages to grow (64KiB in size) and returns the
+    /// previous size of memory, in pages.
+    ///
+    /// [MDN documentation](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/WebAssembly/Memory/grow)
+    #[wasm_bindgen(catch, method, js_namespace = WebAssembly)]
```
```suggestion
    /// previous size of memory, in pages.
    ///
    /// # Reimplementation
    ///
    /// We re-implement `WebAssembly.Memory.grow` because it is
    /// different from what `wasm-bindgen` declares. It marks the function
    /// as `catch`, which means it can throw an exception.
    ///
    /// See [the opened patch](https://github.com/rustwasm/wasm-bindgen/pull/2599).
    ///
    /// # Exceptions
    ///
    /// A `RangeError` is thrown if adding pages would exceed the maximum
    /// memory.
    ///
    /// [MDN documentation](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/WebAssembly/Memory/grow)
    #[wasm_bindgen(catch, method, js_namespace = WebAssembly)]
```

Hywan commented on lib/js-api/src/externals/memory.rs:
```diff
@@ -0,0 +1,319 @@
+    /// modify the memory contents in any way including by calling a wasm
+    /// function that writes to the memory or by resizing the memory.
+    pub unsafe fn data_unchecked(&self) -> &[u8] {
+        unimplemented!("direct data pointer access is not possible in js");
```
```suggestion
        unimplemented!("direct data pointer access is not possible in JavaScript");
```

I would suggest to basically remove the `data_unchecked` function, so that it emits a compiler error to the user, instead of a runtime error.

Hywan commented on lib/js-api/src/externals/memory.rs:
```diff
@@ -0,0 +1,319 @@
+    /// by resizing this Memory.
+    #[allow(clippy::mut_from_ref)]
+    pub unsafe fn data_unchecked_mut(&self) -> &mut [u8] {
+        unimplemented!("direct data pointer access is not possible in js");
```
```suggestion
        unimplemented!("direct data pointer access is not possible in JavaScript");
```

I would suggest to basically remove the `data_unchecked_mut` function, so that it emits a compiler error to the user, instead of a runtime error.

Hywan commented on lib/js-api/src/externals/memory.rs:
```diff
@@ -0,0 +1,319 @@
+
+    /// Returns the pointer to the raw bytes of the `Memory`.
+    pub fn data_ptr(&self) -> *mut u8 {
+        unimplemented!("direct data pointer access is not possible in js");
```
```suggestion
        unimplemented!("direct data pointer access is not possible in JavaScript");
```

I would suggest to basically remove the `data_ptr` function, so that it emits a compiler error to the user, instead of a runtime error.

Hywan commented on lib/js-api/src/externals/memory.rs:
```diff
@@ -0,0 +1,319 @@
+
+    /// Returns the size (in bytes) of the `Memory`.
+    pub fn data_size(&self) -> u64 {
+        let bytes = js_sys::Reflect::get(&self.vm_memory.memory.buffer(), &"byteLength".into())
+            .unwrap()
+            .as_f64()
+            .unwrap() as u64;
+        return bytes;
```
```suggestion
        js_sys::Reflect::get(&self.vm_memory.memory.buffer(), &"byteLength".into())
            .unwrap()
            .as_f64()
            .unwrap() as _
```

Hywan commented on lib/js-api/src/externals/memory.rs:
````diff
@@ -0,0 +1,319 @@
+    /// # }
+    /// ```
+    pub fn view<T: ValueType>(&self) -> MemoryView<T> {
+        unimplemented!();
````
I would suggest to add a comment here.
Or remove the `view` function (see `data_unchecked` etc. above).

syrusakbary replied:
Fixed

Hywan commented on lib/js-api/src/externals/table.rs:
```diff
@@ -0,0 +1,167 @@
+        _src_index: u32,
+        _len: u32,
+    ) -> Result<(), RuntimeError> {
+        unimplemented!();
```
Add a message, or remove the `copy` function entirely.

syrusakbary replied:
Fixed

Hywan commented on lib/js-api/src/instance.rs:
```diff
@@ -0,0 +1,143 @@
+use crate::externals::Extern;
+use crate::module::Module;
+use crate::store::Store;
+// use crate::{HostEnvInitError, LinkError, RuntimeError};
```
```suggestion
```

Hywan commented on lib/js-api/src/instance.rs:
```diff
@@ -0,0 +1,143 @@
+    /// A runtime error occured while invoking the start function
+    #[cfg_attr(feature = "std", error("Start error: {0}"))]
+    Start(String),
+    // /// Error occurred when initializing the host environment.
+    // #[error(transparent)]
+    // HostEnvInitialization(HostEnvInitError),
```
```suggestion
```

Hywan commented on lib/js-api/src/instance.rs:
```diff
@@ -0,0 +1,143 @@
+    // HostEnvInitialization(HostEnvInitError),
+}
+
+// impl From<wasmer_engine::InstantiationError> for InstantiationError {
+//     fn from(other: wasmer_engine::InstantiationError) -> Self {
+//         match other {
+//             wasmer_engine::InstantiationError::Link(e) => Self::Link(e),
+//             wasmer_engine::InstantiationError::Start(e) => Self::Start(e),
+//         }
+//     }
+// }
+
+// impl From<HostEnvInitError> for InstantiationError {
+//     fn from(other: HostEnvInitError) -> Self {
+//         Self::HostEnvInitialization(other)
+//     }
+// }
```
```suggestion
```

Hywan commented on lib/js-api/src/instance.rs:
```diff
@@ -0,0 +1,143 @@
+    ///
+    /// The function can return [`InstantiationError`]s.
+    ///
+    /// Those are, as defined by the spec:
+    ///  * Link errors that happen when plugging the imports into the instance
+    ///  * Runtime errors that happen when running the module `start` function.
```
Nah, those aren't returned if I'm following the code correctly. Can you confirm?

syrusakbary replied:
Let me fix that

syrusakbary replied:
Fixed!

Hywan commented on lib/js-api/src/lib.rs:
```diff
@@ -0,0 +1,147 @@
+    )
+)]
+
+//! This crate contains the `wasmer` API. The `wasmer` API facilitates the efficient,
+//! sandboxed execution of [WebAssembly (Wasm)][wasm] modules.
```
Incorrect. Can you update please?

syrusakbary replied:
What you want to update here? `wasmer` for `wasmer-js`? Or saying that it targets JS?

syrusakbary replied:
Updated

Hywan commented on lib/js-api/src/lib.rs:
```diff
@@ -0,0 +1,147 @@
+pub use crate::types::{Val as Value, ValType as Type};
+pub use crate::utils::is_wasm;
+
+// #[cfg(feature = "experimental-reference-types-extern-ref")]
+// pub use wasmer_types::ExternRef;
```
```suggestion
```

Hywan commented on lib/js-api/src/trap.rs:
```diff
@@ -0,0 +1,167 @@
+    }
+}
+
+// fn _assert_trap_is_sync_and_send(t: &Trap) -> (&dyn Sync, &dyn Send) {
+//     (t, t)
+// }
+
```
```suggestion
```

Hywan commented on lib/js-api/src/types.rs:
```diff
@@ -0,0 +1,46 @@
+        ValType::I64 => Val::I64(js_val.as_f64().unwrap() as _),
+        ValType::F32 => Val::F32(js_val.as_f64().unwrap() as _),
+        ValType::F64 => Val::F64(js_val.as_f64().unwrap()),
+        _ => unimplemented!("The type is not yet supported in the JS Function API"),
```
```suggestion
        t => unimplemented!("The type `{:?}` is not yet supported in the JS Function API", t),
```

Hywan commented on lib/js-api/src/types.rs:
```diff
@@ -0,0 +1,46 @@
+            Self::F32(f) => JsValue::from_f64(*f as f64),
+            Self::F64(f) => JsValue::from_f64(*f),
+            Self::FuncRef(func) => func.as_ref().unwrap().exported.function.clone().into(),
+            _ => unimplemented!("The type is not yet supported in the JS Function API"),
```
```suggestion
            v => unimplemented!("The value `{:?}` is not yet supported in the JS Function API", v),
```

Hywan commented on lib/js-api/tests/externals.rs:
```diff
@@ -0,0 +1,418 @@
+    let store = Store::default();
+    let global_i32 = Global::new(&store, Value::I32(10));
+    assert_eq!(global_i32.get(), Value::I32(10));
+    // let global_i64 = Global::new(&store, Value::I64(20));
+    // assert_eq!(global_i64.get(), Value::I64(20));
```
Why is it commented?

syrusakbary replied:
```
    // 64-bit values are not yet fully supported in some versions of Node
    // Commenting this tests for now:
```

Hywan commented on lib/js-api/tests/externals.rs:
```diff
@@ -0,0 +1,418 @@
+    // assert_eq!(global_i64.get(), Value::I64(20));
+    let global_f32 = Global::new(&store, Value::F32(10.0));
+    assert_eq!(global_f32.get(), Value::F32(10.0));
+    // let global_f64 = Global::new(&store, Value::F64(20.0));
+    // assert_eq!(global_f64.get(), Value::F64(20.0));
```
Why is it commented?

syrusakbary replied:
I added a comment to the code clarifying:
```
    // 64-bit values are not yet fully supported in some versions of Node
    // Commenting this tests for now:
```

Hywan commented on lib/js-api/tests/externals.rs:
```diff
@@ -0,0 +1,418 @@
+//     assert_eq!(*table.ty(), table_type);
+//     let _elem = table.get(0).unwrap();
+//     // assert_eq!(elem.funcref().unwrap(), f);
+//     Ok(())
+// }
+
+// #[test]
+// #[ignore]
+// fn table_set() -> Result<()> {
+//     // Table set not yet tested
+//     Ok(())
+// }
+
+// #[test]
+// fn table_grow() -> Result<()> {
+//     let store = Store::default();
+//     let table_type = TableType {
+//         ty: Type::FuncRef,
+//         minimum: 0,
+//         maximum: Some(10),
+//     };
+//     let f = Function::new_native(&store, |num: i32| num + 1);
+//     let table = Table::new(&store, table_type, Value::FuncRef(Some(f.clone())))?;
+//     // Growing to a bigger maximum should return None
+//     let old_len = table.grow(12, Value::FuncRef(Some(f.clone())));
+//     assert!(old_len.is_err());
+
+//     // Growing to a bigger maximum should return None
+//     let old_len = table.grow(5, Value::FuncRef(Some(f.clone())))?;
+//     assert_eq!(old_len, 0);
+
+//     Ok(())
+// }
+
+// #[test]
+// #[ignore]
+// fn table_copy() -> Result<()> {
+//     // TODO: table copy test not yet implemented
+//     Ok(())
+// }
```
Why is it commented?

syrusakbary replied:
Commented in the code

Hywan commented on lib/js-api/tests/externals.rs:
```diff
@@ -0,0 +1,418 @@
+            attempted_delta: 10.into()
+        })
+    );
+
+    // let bad_desc = MemoryType::new(Pages(15), Some(Pages(10)), false);
+    // let bad_result = Memory::new(&store, bad_desc);
+
+    // // assert!(matches!(bad_result, Err(MemoryError::InvalidMemory { .. })));
+    // assert!(bad_result.is_err());
```
```suggestion
```

Hywan reviewed (changes requested):
1. Did you try to compile this new crate in `no_std` mode?

Other than my suggestions and comments, I think the PR looks good!

I'm still not convinced by the name. See our private Slack discussion.

Hywan commented on lib/js-api/README.md:
```diff
@@ -0,0 +1,75 @@
+    let instance = Instance::new(&module, &import_object).unwrap();
+
+    let add_one = instance.exports.get_function("add_one").unwrap();
+    let result = add_one.call(&[Value::I32(42)])?;
```
```suggestion
    let result = add_one.call(&[Value::I32(42)]).unwrap();
```

`do_add_one_in_wasmer` doesn't return a `Result`.

syrusakbary replied:
Good call

syrusakbary commented on lib/js-api/Cargo.toml:
```diff
@@ -0,0 +1,46 @@
+# `wasm-opt` is on by default in for the release profile, but it can be
+# disabled by setting it to `false`
+[package.metadata.wasm-pack.profile.release]
+wasm-opt = false
```
In macOS M1 doesn't work. This metadata has no effect on how people use or target the package

syrusakbary commented on lib/js-api/README.md:
````diff
@@ -0,0 +1,75 @@
+    assert_eq!(result[0], Value::I32(43));
+    result[0].unwrap_i32()
+}
+```
````
Is the simplest one I could find to showcase the project. It's similar in simplicity to the one in wasmer/api, which I really think is ideal:

https://github.com/wasmerio/wasmer/tree/master/lib/api

syrusakbary commented on lib/vm/src/lib.rs:
```diff
@@ -48,7 +47,6 @@ pub use crate::instance::{
 };
 pub use crate::memory::{LinearMemory, Memory, MemoryError, MemoryStyle};
 pub use crate::mmap::Mmap;
-pub use crate::module::{ExportsIterator, ImportsIterator, ModuleInfo};
```
Sure

syrusakbary replied:
Fixed

syrusakbary commented:
bors try

syrusakbary commented:
> Did you try to compile this new crate in no_std mode?
Yes, I did... and it compiles properly :) (just did the last adjustments)

Hywan commented on lib/vm/src/lib.rs:
```diff
@@ -62,6 +60,11 @@ pub use crate::vmcontext::{
 pub use crate::vmoffsets::{TargetSharedSignatureIndex, VMOffsets};
 use loupe::MemoryUsage;
 pub use wasmer_types::VMExternRef;
+#[deprecated(
+    since = "2.1.0",
+    note = "ModuleInfo, ExportsIterator, ImportsIterator should be imported from wasmer_types."
+)]
```
I'm not sure why we should make it deprecated. It's fine to re-export types from another crate here. What do you think?

Hywan commented on lib/js-api/src/externals/memory.rs:
````diff
@@ -0,0 +1,331 @@
+    /// }
+    /// # }
+    /// ```
+    pub fn view<T: ValueType>(&self) -> MemoryView<T> {
+        unimplemented!("The view function is not yet implemented in Wasmer Javascript");
+    }
+
+    /// example view
+    pub fn uint8view(&self) -> js_sys::Uint8Array {
+        js_sys::Uint8Array::new(&self.vm_memory.memory.buffer())
+    }
````
`uint8view` is an example here, not a real method it seems like.

However, we can improve the `view` method to produce a `Uint8Array` or simply to provide an `ArrayBuffer`. We could even mark `view` as unimplemented, and provide a `buffer` method. Or just re-use `data_ptr` to provide an `ArrayBuffer`.

I think we should try to make `view` working with `ArrayBuffer`, that would be the best.

Hywan commented:
It is possible to merge `wasmer-js` into `wasmer-api`, which makes a lot of sense, as discussed in private in Slack. Here is the PR, https://github.com/wasmerio/wasmer/pull/2489/files.

syrusakbary commented:
bors r+
